crates/ra_env/src/lib.rs

```diff
@@ -0,0 +1,70 @@
+//! This crate contains a single public function
+//! [`get_path_for_executable`](fn.get_path_for_executable.html).
+//! See docs there for more information.
+
+use anyhow::{Error, Result};
+use std::env;
+use std::path::Path;
+use std::process::Command;
+
+/// Return a `String` to use for the given executable.
+///
+/// E.g., `get_path_for_executable("cargo")` may return just `cargo` if that
+/// gives a valid Cargo executable; or it may return a full path to a valid
+/// Cargo.
+pub fn get_path_for_executable(executable_name: impl AsRef<str>) -> Result<String> {
+    // The current implementation checks three places for an executable to use:
+    // 1) Appropriate environment variable (erroring if this is set but not a usable executable)
+    //      example: for cargo, this checks $CARGO environment variable; for rustc, $RUSTC; etc
+    // 2) `<executable_name>`
+    //      example: for cargo, this tries just `cargo`, which will succeed if `cargo` is on the $PATH
+    // 3) `~/.cargo/bin/<executable_name>`
+    //      example: for cargo, this tries ~/.cargo/bin/cargo
+    //      It seems that this is a reasonable place to try for cargo, rustc, and rustup
+    let executable_name = executable_name.as_ref();
+    let env_var = executable_name.to_ascii_uppercase();
+    if let Ok(path) = env::var(&env_var) {
+        if is_valid_executable(&path) {
+            Ok(path)
+        } else {
+            Err(Error::msg(format!(
+                "`{}` environment variable points to something that's not a valid executable",
+                env_var
+            )))
+        }
+    } else {
+        let final_path: Option<String> = if is_valid_executable(executable_name) {
+            Some(executable_name.to_owned())
+        } else {
+            if let Some(mut path) = dirs::home_dir() {
+                path.push(".cargo");
+                path.push("bin");
+                path.push(executable_name);
+                if is_valid_executable(&path) {
+                    Some(path.into_os_string().into_string().expect("Invalid Unicode in path"))
+                } else {
+                    None
+                }
+            } else {
+                None
+            }
+        };
```

Comment:
I think this code can benefit from early returns to simplify it.

Reply:
Will make the change.  I know some prefer not to use early returns, but if people are fine with them here I agree that they will simplify this.

Reply:
Fixed in f64740b